blockchain/sizehelper.go

```diff
@@ -63,14 +63,25 @@ const (
 	loadFactorNum = 13
 	loadFactorDen = 2
 
-	// maxAlloc is the maximum size of an allocation. On 64-bit,
-	// it's theoretically possible to allocate 1<<heapAddrBits bytes. On
-	// 32-bit, however, this is one less than 1<<32 because the
-	// number of bytes in the address space doesn't actually fit
-	// in a uintptr.
+	// _64bit = 1 on 64-bit systems, 0 on 32-bit systems
+	_64bit = 1 << (^uintptr(0) >> 63) / 2
+
+	// PtrSize is the size of a pointer in bytes - unsafe.Sizeof(uintptr(0))
+	// but as an ideal constant. It is also the size of the machine's native
+	// word size (that is, 4 on 32-bit systems, 8 on 64-bit).
+	PtrSize = 4 << (^uintptr(0) >> 63)
+
+	// heapAddrBits is the number of bits in a heap address that's actually
+	// available for memory allocation.
 	//
-	// NOTE (kcalvinalvin): I just took the constant for a 64 bit system.
-	maxAlloc = 281474976710656
+	// NOTE (guggero): For 64-bit systems, we just assume 40 bits of address
+	// space available, as that seems to be the lowest common denominator.
+	// See heapAddrBits in runtime/malloc.go of the standard library for
+	// more details
+	heapAddrBits = 32 + (_64bit * 8)
```

Comment:
@kcalvinalvin: If we want to restore the original functionality, we can just set this to `32 + (_64bit * 32)`. But I'm not sure what exactly how `calculateRoughMapSize()` is used in the first place, whether this needs to represent the absolute maximum or just a reasonable maximum we don't want to go over in any case?

Reply:
Hmm the problem is that `sizehelper.go` would be completely broken for a 32 bit system.

`calculateRoughMapSize()` is used to get how much the go map is using for the utxocache. The problem would be that it's specifically coded for 64 bit systems and wouldn't return the correct values for a 32 bit system. This would mean that it's inaccurate in calculating how much memory the map is using and the cache may be using more than what the user designated.

I don't think there's an easy fix here other than disabling the utxo cache for 32 bit systems...

Reply:
Okay, I guess that's something to look into. But to fix the _compilation_ on 32-bit systems, what would you propose? I tried to make the code more similar to `runtime/malloc.go` with my changes. But I don't think I understand what the impact of `maxAlloc` is on the overall correctness of `calculateRoughMapSize()`.

Reply:
Ah `maxAlloc` is just to check if whatever the user requested was over that size. If it was over, then the runtime won't try to allocate that much space and will just try default to 0 `make([]uint64, 0)` <- this 0 here.

```
func Test1(t *testing.T) {
        mib := 300_000
        hint := calculateMinEntries(mib*1024*1204, bucketSize+avgEntrySize) - 1
        size := calculateRoughMapSize(hint, bucketSize)
        fmt.Println(size, hint)
}
```

Played around with that code above and seems like the new `maxAlloc` will cause issues when the user allocates ~400GiB as the utxocache when the old `maxAlloc` value will allow for that. The new one is ok at ~300GiB.

Seems ok since we're not expecting anyone to do that soon? Maybe just a comment on it explaining this would be good.

Reply:
Okay, cool. Thanks for verifying. Added the comment in the latest push, what do you think?

Reply:
Looks great!